File: ios/Classes/EventChannel/ReceiveMessageStreamHandler.swift
//
//  ReceiveMessageStreamHandler.swift
//  flutter_libpd
//
//  Created by flav on 01/06/2022.
//

import Foundation

public class ReceiveMessageStreamHandler: NSObject, FlutterStreamHandler {
    private let messageReceiver: PdMessageReceiver
    
    public init(messageReceiver: PdMessageReceiver) {
        self.messageReceiver = messageReceiver
    }
    
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        guard let source = makeSource(arguments) else { return makeError(arguments) }
        messageReceiver.addListener(source: source) { message in
            events(["source": message.source,
                    "symbol": message.symbol,
                    "value": message.arguments] as [String: Any?])
        }
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        guard let source = makeSource(arguments) else { return makeError(arguments) }
        messageReceiver.removeListener(source: source)
        return nil
    }
    
    private func makeSource(_ arguments: Any?) -> String? {
        guard let arguments = arguments as? [String], arguments.count == 1,
              let source = arguments.first, !source.isEmpty else {
                  return nil
              }
        return source
    }
    
    private func makeError(_ arguments: Any?) -> FlutterError {
        FlutterError(code: "INVALID_PARAMETERS", message: "Expected param: [String] of size 1", details: arguments)
    }
}

File: android/src/main/kotlin/com/ovaom/flutter_libpd/event_channel/ReceiveMessageStreamHandler.kt
package com.ovaom.flutter_libpd.event_channel

import com.ovaom.flutter_libpd.pd_controller.pd_message_receiver.PdMessageReceiver
import io.flutter.plugin.common.EventChannel.EventSink
import io.flutter.plugin.common.EventChannel.StreamHandler

class ReceiveMessageStreamHandler(private val messageReceiver: PdMessageReceiver) : StreamHandler {

    override fun onListen(arguments: Any?, events: EventSink) {
        val source = makeSource(arguments) ?: return deliverInvalidParamsError(arguments, events)
        messageReceiver.addListener(source) {
            events.success(
                mapOf(
                    "source" to it.source,
                    "symbol" to it.symbol,
                    "value" to it.arguments
                )
            )
        }
    }

    override fun onCancel(arguments: Any?) {
        makeSource(arguments)?.let {
            messageReceiver.removeListener(it)
        }
    }

    private fun deliverInvalidParamsError(arguments: Any?, events: EventSink) {
        events.error("INVALID_PARAMETERS", "Expected param: [String] of size 1", arguments)
    }

    private fun makeSource(arguments: Any?): String? {
        (arguments as? List<*>)?.filterIsInstance<String>()?.let {
            if (it.size == 1 && it[0].isNotEmpty()) { return it[0] }
        }
        return null
    }
}